File: dp/0837-new-21-game.py
from typing import List
import unittest

# https://leetcode.com/problems/new-21-game/

# python3 -m unittest dp/0837-new-21-game.py

# In short, you're tasked with finding the likelihood of her final score being between k and n.


class Solution(unittest.TestCase):
    # # Approach #1: Top-Down Dynamic Programming
    # # Time: O(k * maxPts)
    # # Space: O(k)
    # def new21Game(self, n: int, k: int, maxPts: int) -> float:
    #     memo: List[float] = [-1] * k

    #     def dp(points: int) -> float:
    #         if points >= k:
    #             return 1 if points <= n else 0
    #         if memo[points] != -1:
    #             return memo[points]
    #         score = 0
    #         for point in range(1, maxPts + 1):
    #             score += dp(points + point)
    #         memo[points] = score / maxPts
    #         return memo[points]

    #     return dp(0)

    # # Approach #2: Bottom-Up Dynamic Programming
    # # Time: O(n * maxPts)
    # # Space: O(n)
    # def new21Game(self, n: int, k: int, maxPts: int) -> float:
    #     dp: List[float] = [0] * (n + 1)
    #     dp[0] = 1
    #     for i in range(1, n + 1):
    #         for j in range(1, maxPts + 1):
    #             if i - j >= 0 and i - j < k:
    #                 dp[i] += dp[i - j] / maxPts
    #     return sum(dp[k:])

    # # Approach #3: Bottom-Up Dynamic Programming (Optimized)
    # # Time: O(n)
    # # Space: O(n)
    # def new21Game(self, n: int, k: int, maxPts: int) -> float:
    #     dp: List[float] = [0] * (n + 1)
    #     dp[0] = 1
    #     wsum = 1 if k > 0 else 0
    #     for i in range(1, n + 1):
    #         dp[i] = wsum / maxPts
    #         if i < k:
    #             wsum += dp[i]
    #         if i - maxPts >= 0 and i - maxPts < k:
    #             wsum -= dp[i - maxPts]
    #     return sum(dp[k:])

    # Approach #4: Bottom-Up Dynamic Programming (Optimized)
    # Time: O(k + maxPts)
    # Space: O(k + maxPts)
    def new21Game(self, n: int, k: int, maxPts: int) -> float:
        if k == 0 or k + maxPts <= n:
            return 1
        dp: List[float] = [0] * (k + maxPts)
        wsum = 0
        for i in range(k, k + maxPts):
            if i <= n:
                dp[i] = 1
                wsum += 1
        for i in range(k - 1, -1, -1):
            dp[i] = wsum / maxPts
            wsum += dp[i] - dp[i + maxPts]
        return dp[0]

    def test(self):
        for n, k, maxPts, expected in [
            (10, 1, 10, 1.00000),
            (6, 1, 10, 0.60000),
            (21, 17, 10, 0.73278),
        ]:
            output = self.new21Game(n, k, maxPts)
            output = round(output, 5)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: dp/0837-new-21-game.go
package dp

// In short, you're tasked with finding the likelihood of her final score being between k and n.

// https://leetcode.com/problems/new-21-game/

// Approach #4: Bottom-Up Dynamic Programming (Optimized)
// Time: O(k + maxPts)
// Space: O(k + maxPts)
func new21Game(n int, k int, maxPts int) float64 {
	if k == 0 || k+maxPts <= n {
		return 1
	}

	dp := make([]float64, k+maxPts)

	var wsum float64

	for i := k; i < k+maxPts; i++ {
		if i <= n {
			dp[i] = 1
			wsum++
		}
	}

	for i := k - 1; i >= 0; i-- {
		dp[i] = wsum / float64(maxPts)
		wsum += dp[i] - dp[i+maxPts]
	}

	return dp[0]
}

// // Approach #3: Bottom-Up Dynamic Programming (Optimized)
// // Time: O(n)
// // Space: O(n)
// func new21Game(n int, k int, maxPts int) float64 {
// 	dp := make([]float64, n+1)
// 	dp[0] = 1

// 	var wsum float64
// 	if k > 0 {
// 		wsum = 1
// 	}

// 	for i := 1; i <= n; i++ {
// 		dp[i] = wsum / float64(maxPts)
// 		if i < k {
// 			wsum += dp[i]
// 		}

// 		if i-maxPts >= 0 && i-maxPts < k {
// 			wsum -= dp[i-maxPts]
// 		}
// 	}

// 	var score float64
// 	for i := k; i <= n; i++ {
// 		score += dp[i]
// 	}

// 	return score
// }

// // Approach #2: Bottom-Up Dynamic Programming
// // Time: O(n * maxPts)
// // Space: O(n)
// func new21Game(n int, k int, maxPts int) float64 {
// 	dp := make([]float64, n+1)
// 	dp[0] = 1

// 	for i := 1; i <= n; i++ {
// 		for j := 1; j <= maxPts; j++ {
// 			if i-j >= 0 && i-j < k {
// 				dp[i] += dp[i-j] / float64(maxPts)
// 			}
// 		}
// 	}

// 	var score float64
// 	for i := k; i <= n; i++ {
// 		score += dp[i]
// 	}

// 	return score
// }

// // Approach #1: Top-Down Dynamic Programming
// // Time: O(k * maxPts)
// // Space: O(k)
// func new21Game(n int, k int, maxPts int) float64 {
// 	memo := make([]*float64, k)

// 	var dp func(int) float64

// 	dp = func(points int) float64 {
// 		if points >= k {
// 			if points <= n {
// 				return 1
// 			}

// 			return 0
// 		}

// 		if memo[points] != nil {
// 			return *memo[points]
// 		}

// 		var score float64
// 		for point := 1; point <= maxPts; point++ {
// 			score += dp(points + point)
// 		}

// 		score /= float64(maxPts)
// 		memo[points] = &score

// 		return score
// 	}

// 	return dp(0)
// }
